File: kirun-js/src/engine/dsl/transformer/ASTToJSON.ts
import {
    ArgumentNode,
    ComplexValueNode,
    EventDeclNode,
    ExpressionNode,
    FunctionDefNode,
    ParameterDeclNode,
    SchemaLiteralNode,
    StatementNode,
} from '../parser/ast';
import { SchemaTransformer } from './SchemaTransformer';

/**
 * AST to JSON Transformer
 * Converts parsed AST to FunctionDefinition JSON
 */
export class ASTToJSONTransformer {
    /**
     * Transform AST to FunctionDefinition JSON
     * Only includes non-empty/non-default fields
     */
    public transform(ast: FunctionDefNode): any {
        const json: any = {
            name: ast.name,
        };

        // Only add namespace if not empty
        if (ast.namespace) {
            json.namespace = ast.namespace;
        }

        // Only add parameters if not empty
        const parameters = this.transformParameters(ast.parameters);
        if (Object.keys(parameters).length > 0) {
            json.parameters = parameters;
        }

        // Only add events if not empty
        const events = this.transformEvents(ast.events);
        if (Object.keys(events).length > 0) {
            json.events = events;
        }

        // Always add steps
        json.steps = this.transformSteps(ast.logic);

        return json;
    }

    /**
     * Transform parameters
     */
    private transformParameters(params: ParameterDeclNode[]): any {
        const result: any = {};
        for (const param of params) {
            result[param.name] = {
                parameterName: param.name,
                schema: SchemaTransformer.transform(param.schema.schemaSpec),
                variableArgument: false,
                type: 'EXPRESSION',
            };
        }
        return result;
    }

    /**
     * Transform events
     */
    private transformEvents(events: EventDeclNode[]): any {
        const result: any = {};
        for (const event of events) {
            result[event.name] = {
                name: event.name,
                parameters: this.transformEventParameters(event.parameters),
            };
        }
        return result;
    }

    /**
     * Transform event parameters (these are schemas, not Parameter objects)
     */
    private transformEventParameters(params: ParameterDeclNode[]): any {
        const result: any = {};
        for (const param of params) {
            result[param.name] = SchemaTransformer.transform(param.schema.schemaSpec);
        }
        return result;
    }

    /**
     * Transform statements (top-level and nested)
     */
    private transformSteps(statements: StatementNode[]): any {
        const result: any = {};
        for (const stmt of statements) {
            result[stmt.statementName] = this.transformStatement(stmt);
        }
        return result;
    }

    /**
     * Transform single statement
     * Only includes non-empty/non-default fields
     */
    private transformStatement(stmt: StatementNode): any {
        const json: any = {
            statementName: stmt.statementName,
            namespace: stmt.functionCall.namespace,
            name: stmt.functionCall.name,
        };

        // Only add parameterMap if not empty
        const parameterMap = this.transformParameterMap(stmt.functionCall.argumentsMap);
        if (Object.keys(parameterMap).length > 0) {
            json.parameterMap = parameterMap;
        }

        // Only add dependentStatements if not empty
        const dependentStatements = this.createDependentStatementsMap(stmt);
        if (Object.keys(dependentStatements).length > 0) {
            json.dependentStatements = dependentStatements;
        }

        // Only add executeIftrue if not empty
        const executeIftrue = this.createExecuteIfMap(stmt.executeIfSteps);
        if (Object.keys(executeIftrue).length > 0) {
            json.executeIftrue = executeIftrue;
        }

        // Only add comment if not empty
        if (stmt.comment && stmt.comment.trim()) {
            json.comment = stmt.comment;
        }

        return json;
    }

    /**
     * Create dependentStatements map from AFTER clause and nested blocks
     */
    private createDependentStatementsMap(stmt: StatementNode): any {
        const result: any = {};

        // Add explicit AFTER dependencies
        for (const stepRef of stmt.afterSteps) {
            result[stepRef] = true; // true = depends on this step
        }

        // Add implicit dependencies from nested blocks
        // Nested statements depend on their parent's specific events
        for (const [blockName, nestedStmts] of stmt.nestedBlocks) {
            // Create dependency key based on block name
            // e.g., "Steps.loop.iteration", "Steps.if.true"
            const dependencyKey = `Steps.${stmt.statementName}.${blockName}`;

            // Mark nested statements as depending on this block
            // (These will be added when we transform nested statements)
        }

        return result;
    }

    /**
     * Create executeIftrue map from IF clause
     */
    private createExecuteIfMap(executeIfSteps: string[]): any {
        const result: any = {};
        for (const stepRef of executeIfSteps) {
            result[stepRef] = true;
        }
        return result;
    }

    /**
     * Transform parameter map (function arguments)
     * Skips parameters with empty values
     * Supports multi-value parameters
     */
    private transformParameterMap(argsMap: Map<string, ArgumentNode>): any {
        const result: any = {};

        for (const [paramName, arg] of argsMap) {
            // Skip parameters with empty values
            if (this.isEmptyArgument(arg)) {
                continue;
            }

            result[paramName] = {};

            // Handle multi-value parameters
            if (arg.isMultiValue()) {
                for (let i = 0; i < arg.values.length; i++) {
                    const paramRef = this.transformArgumentValue(arg.values[i], i + 1);
                    result[paramName][paramRef.key] = paramRef;
                }
            } else {
                const paramRef = this.transformArgumentValue(arg.value, 1);
                result[paramName][paramRef.key] = paramRef;
            }
        }

        return result;
    }

    /**
     * Check if an argument is empty (no value provided)
     * Note: Empty expressions and undefined values are valid and should NOT be filtered out
     * Only filter out parser artifacts like stray delimiters
     */
    private isEmptyArgument(arg: ArgumentNode): boolean {
        if (arg.value instanceof ExpressionNode) {
            const expr = arg.value.expressionText.trim();
            // Only filter out parser artifacts (stray delimiters)
            // Empty expressions ('') are valid and should be preserved
            return expr === ',' || expr === ')';
        }
        // Both null and undefined are valid values and should be preserved
        // ComplexValueNode is never "empty" - it always has a value (even if null/undefined)
        return false;
    }

    /**
     * Transform single argument value to ParameterReference
     * @param value The argument value (expression, complex value, or schema literal)
     * @param order The order index for multi-value parameters (1-based)
     */
    private transformArgumentValue(
        value: ExpressionNode | ComplexValueNode | SchemaLiteralNode,
        order: number,
    ): any {
        const key = this.generateUUID();

        if (value instanceof ExpressionNode) {
            return {
                key,
                type: 'EXPRESSION',
                expression: value.expressionText,
                value: undefined,
                order: order,
            };
        } else if (value instanceof ComplexValueNode) {
            return {
                key,
                type: 'VALUE',
                value: value.value,
                expression: undefined,
                order: order,
            };
        } else if (value instanceof SchemaLiteralNode) {
            // Schema literal with optional default value
            const schema = SchemaTransformer.transform(value.schema.schemaSpec);

            if (value.defaultValue) {
                // Has default value - evaluate it
                return {
                    key,
                    type: 'VALUE',
                    value: this.evaluateDefaultValue(value.defaultValue, schema),
                    expression: undefined,
                    order: order,
                };
            } else {
                // No default value - just pass the schema
                return {
                    key,
                    type: 'VALUE',
                    value: schema,
                    expression: undefined,
                    order: order,
                };
            }
        }

        // Fallback
        return {
            key,
            type: 'VALUE',
            value: null,
            expression: undefined,
            order: order,
        };
    }

    /**
     * Evaluate default value for schema literal
     */
    private evaluateDefaultValue(defaultValueExpr: ExpressionNode, schema: any): any {
        const exprText = defaultValueExpr.expressionText.trim();

        // Try to parse as literal value
        if (exprText === '[]') {
            return [];
        }
        if (exprText === '{}') {
            return {};
        }
        if (exprText === 'null') {
            return null;
        }
        if (exprText === 'true') {
            return true;
        }
        if (exprText === 'false') {
            return false;
        }

        // Try to parse as number
        const num = parseFloat(exprText);
        if (!isNaN(num)) {
            return num;
        }

        // Try to parse as string literal
        if (
            (exprText.startsWith('"') && exprText.endsWith('"')) ||
            (exprText.startsWith("'") && exprText.endsWith("'"))
        ) {
            return exprText.slice(1, -1);
        }

        // Otherwise, return as-is (might be an expression)
        return exprText;
    }

    /**
     * Generate UUID for parameter reference keys
     */
    private generateUUID(): string {
        // Simple UUID v4 generation
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
            const r = (Math.random() * 16) | 0;
            const v = c === 'x' ? r : (r & 0x3) | 0x8;
            return v.toString(16);
        });
    }

    /**
     * Flatten nested blocks into steps
     * This adds all nested statements to the top-level steps object
     * and sets their dependentStatements appropriately
     */
    public flattenNestedBlocks(ast: FunctionDefNode): void {
        const allStatements: StatementNode[] = [...ast.logic];

        // Process each top-level statement
        for (const stmt of ast.logic) {
            this.collectNestedStatements(stmt, allStatements);
        }

        // Update ast.logic with all statements (flattened)
        ast.logic = allStatements;
    }

    /**
     * Recursively collect nested statements
     * NOTE: Nesting is implicit (derived from expression analysis)
     * We do NOT add block dependencies to dependentStatements
     * Only explicit AFTER clauses become dependentStatements
     */
    private collectNestedStatements(stmt: StatementNode, allStatements: StatementNode[]): void {
        for (const [, nestedStmts] of stmt.nestedBlocks) {
            for (const nestedStmt of nestedStmts) {
                // Do NOT add implicit block dependency - nesting is derived from expressions
                // Only explicit AFTER clauses should become dependentStatements

                // Add to all statements
                allStatements.push(nestedStmt);

                // Recursively process this statement's nested blocks
                this.collectNestedStatements(nestedStmt, allStatements);
            }
        }

        // Clear nested blocks after flattening
        stmt.nestedBlocks.clear();
    }
}

File: kirun-py/src/kirun_py/dsl/transformer/ast_to_json.py
from __future__ import annotations

import uuid
from typing import Any, Dict, List, Optional

from kirun_py.dsl.parser.ast.argument_node import ArgumentNode, ArgumentValue
from kirun_py.dsl.parser.ast.complex_value_node import ComplexValueNode
from kirun_py.dsl.parser.ast.event_decl_node import EventDeclNode
from kirun_py.dsl.parser.ast.expression_node import ExpressionNode
from kirun_py.dsl.parser.ast.function_def_node import FunctionDefNode
from kirun_py.dsl.parser.ast.parameter_decl_node import ParameterDeclNode
from kirun_py.dsl.parser.ast.schema_literal_node import SchemaLiteralNode
from kirun_py.dsl.parser.ast.statement_node import StatementNode
from kirun_py.dsl.transformer.schema_transformer import SchemaTransformer


class ASTToJSONTransformer:
    """
    AST to JSON Transformer.
    Converts parsed AST to FunctionDefinition JSON.
    """

    def transform(self, ast: FunctionDefNode) -> Any:
        """
        Transform AST to FunctionDefinition JSON.
        Only includes non-empty/non-default fields.
        """
        json_obj: dict = {
            'name': ast.name,
        }

        if ast.namespace:
            json_obj['namespace'] = ast.namespace

        parameters = self._transform_parameters(ast.parameters)
        if parameters:
            json_obj['parameters'] = parameters

        events = self._transform_events(ast.events)
        if events:
            json_obj['events'] = events

        json_obj['steps'] = self._transform_steps(ast.logic)

        return json_obj

    def _transform_parameters(self, params: List[ParameterDeclNode]) -> dict:
        """Transform parameters."""
        result: dict = {}
        for param in params:
            result[param.name] = {
                'parameterName': param.name,
                'schema': SchemaTransformer.transform(param.schema.schema_spec),
                'variableArgument': False,
                'type': 'EXPRESSION',
            }
        return result

    def _transform_events(self, events: List[EventDeclNode]) -> dict:
        """Transform events."""
        result: dict = {}
        for event in events:
            result[event.name] = {
                'name': event.name,
                'parameters': self._transform_event_parameters(event.parameters),
            }
        return result

    def _transform_event_parameters(self, params: List[ParameterDeclNode]) -> dict:
        """Transform event parameters (these are schemas, not Parameter objects)."""
        result: dict = {}
        for param in params:
            result[param.name] = SchemaTransformer.transform(param.schema.schema_spec)
        return result

    def _transform_steps(self, statements: List[StatementNode]) -> dict:
        """Transform statements (top-level and nested)."""
        result: dict = {}
        for stmt in statements:
            result[stmt.statement_name] = self._transform_statement(stmt)
        return result

    def _transform_statement(self, stmt: StatementNode) -> dict:
        """
        Transform single statement.
        Only includes non-empty/non-default fields.
        """
        json_obj: dict = {
            'statementName': stmt.statement_name,
            'namespace': stmt.function_call.namespace,
            'name': stmt.function_call.name,
        }

        parameter_map = self._transform_parameter_map(stmt.function_call.arguments_map)
        if parameter_map:
            json_obj['parameterMap'] = parameter_map

        dependent_statements = self._create_dependent_statements_map(stmt)
        if dependent_statements:
            json_obj['dependentStatements'] = dependent_statements

        execute_if_true = self._create_execute_if_map(stmt.execute_if_steps)
        if execute_if_true:
            json_obj['executeIftrue'] = execute_if_true

        if stmt.comment and stmt.comment.strip():
            json_obj['comment'] = stmt.comment

        return json_obj

    def _create_dependent_statements_map(self, stmt: StatementNode) -> dict:
        """Create dependentStatements map from AFTER clause and nested blocks."""
        result: dict = {}

        for step_ref in stmt.after_steps:
            result[step_ref] = True

        return result

    def _create_execute_if_map(self, execute_if_steps: List[str]) -> dict:
        """Create executeIftrue map from IF clause."""
        result: dict = {}
        for step_ref in execute_if_steps:
            result[step_ref] = True
        return result

    def _transform_parameter_map(self, args_map: Dict[str, ArgumentNode]) -> dict:
        """
        Transform parameter map (function arguments).
        Skips parameters with empty values.
        Supports multi-value parameters.
        """
        result: dict = {}

        for param_name, arg in args_map.items():
            if self._is_empty_argument(arg):
                continue

            result[param_name] = {}

            if arg.is_multi_value():
                for i, val in enumerate(arg.values):
                    param_ref = self._transform_argument_value(val, i + 1)
                    result[param_name][param_ref['key']] = param_ref
            else:
                param_ref = self._transform_argument_value(arg.value, 1)
                result[param_name][param_ref['key']] = param_ref

        return result

    def _is_empty_argument(self, arg: ArgumentNode) -> bool:
        """
        Check if an argument is empty (no value provided).
        Only filter out parser artifacts like stray delimiters.
        """
        if isinstance(arg.value, ExpressionNode):
            expr = arg.value.expression_text.strip()
            return expr in (',', ')')
        return False

    def _transform_argument_value(self, value: ArgumentValue, order: int) -> dict:
        """
        Transform single argument value to ParameterReference.

        Args:
            value: The argument value (expression, complex value, or schema literal).
            order: The order index for multi-value parameters (1-based).
        """
        key = self._generate_uuid()

        if isinstance(value, ExpressionNode):
            return {
                'key': key,
                'type': 'EXPRESSION',
                'expression': value.expression_text,
                'value': None,
                'order': order,
            }
        elif isinstance(value, ComplexValueNode):
            return {
                'key': key,
                'type': 'VALUE',
                'value': value.value,
                'expression': None,
                'order': order,
            }
        elif isinstance(value, SchemaLiteralNode):
            schema = SchemaTransformer.transform(value.schema.schema_spec)

            if value.default_value is not None:
                return {
                    'key': key,
                    'type': 'VALUE',
                    'value': self._evaluate_default_value(value.default_value, schema),
                    'expression': None,
                    'order': order,
                }
            else:
                return {
                    'key': key,
                    'type': 'VALUE',
                    'value': schema,
                    'expression': None,
                    'order': order,
                }

        return {
            'key': key,
            'type': 'VALUE',
            'value': None,
            'expression': None,
            'order': order,
        }

    def _evaluate_default_value(self, default_value_expr: ExpressionNode, schema: Any) -> Any:
        """Evaluate default value for schema literal."""
        expr_text = default_value_expr.expression_text.strip()

        if expr_text == '[]':
            return []
        if expr_text == '{}':
            return {}
        if expr_text == 'null':
            return None
        if expr_text == 'true':
            return True
        if expr_text == 'false':
            return False

        try:
            num = float(expr_text)
            return num
        except (ValueError, TypeError):
            pass

        if (expr_text.startswith('"') and expr_text.endswith('"')) or \
           (expr_text.startswith("'") and expr_text.endswith("'")):
            return expr_text[1:-1]

        return expr_text

    @staticmethod
    def _generate_uuid() -> str:
        """Generate UUID for parameter reference keys."""
        return str(uuid.uuid4())

    def flatten_nested_blocks(self, ast: FunctionDefNode) -> None:
        """
        Flatten nested blocks into steps.
        Adds all nested statements to the top-level steps object
        and sets their dependentStatements appropriately.
        """
        all_statements: List[StatementNode] = list(ast.logic)

        for stmt in list(ast.logic):
            self._collect_nested_statements(stmt, all_statements)

        ast.logic = all_statements

    def _collect_nested_statements(
        self, stmt: StatementNode, all_statements: List[StatementNode]
    ) -> None:
        """
        Recursively collect nested statements.
        Nesting is implicit (derived from expression analysis).
        Only explicit AFTER clauses become dependentStatements.
        """
        for _block_name, nested_stmts in stmt.nested_blocks.items():
            for nested_stmt in nested_stmts:
                all_statements.append(nested_stmt)
                self._collect_nested_statements(nested_stmt, all_statements)

        stmt.nested_blocks.clear()
